code/cdw.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+CODE DESIGN WORKSHOP 2021
+Functions for pipeline steps
+
+@author:  jenni.saaristo@helsinki.fi
+@version: 2021-04-28
+@notes:   Redo after feedback
+"""
+
+from scipy import signal, stats
+from scipy.spatial.distance import pdist, squareform, cdist
+import numpy as np
+import pandas as pd
+import nibabel as nib
+from itertools import compress
+
+
+def preprocess(bold, labels):
+    """
+    Minimal preprocessing of BOLD data
+    
+    Given a 4D BOLD-data detrends and z-scores it
+    run-wise ("chunks" in labels).
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original BOLD data.
+    labels : DataFrame
+        Labels for MRI frames.
+
+    Returns
+    -------
+    bold_prep : Nifti1Image
+        Preprocessed data.
+    """
+    
+    print('Copying data...')
+    prep = bold.get_fdata()
+    
+    print('Done. Begin preprocessing...')
+    for c in labels.chunks.unique():
+        print('Processing chunk ' + str(c))
+        inds = labels.query('chunks == @c').index
+        this_prep = prep[:,:,:,inds]
+        this_prep = signal.detrend(this_prep)
+        this_prep = stats.zscore(this_prep, axis=-1)
+        
+        # scipy z-score does not watch out for zero-division, so
+        # we have to correct those ourselves. Replacing nans with
+        # zeros makes sense as long as the input is all zeros, too.
+        if prep[:,:,:,inds][np.isnan(this_prep)].sum() == 0:
```

Comment:
while this does work, this line of code is not a direct translation of your intention, recommend using NumPy's `any` function instead of summing boolean values:

```suggestion
        if np.any(this_prep[np.isnan(this_prep)] != 0):
```

Reply:
This actually doesn't do what it's supposed to... changed it to this, though:
`if np.all(prep[:,:,:,inds][np.isnan(this_prep)] == 0):`